### ISSUE TITLE
Append operators for MapProperty, ListProperty, and SetProperty are discarding convention values

### ISSUE THREAD

**Issue by steffenyount** [gradle/gradle#18352]
Calling MapProperty.put()/putAll() or AbstractCollectionProperty.add()/addAll() will discard any convention values I may have set on the MapProperty or AbstractCollectionProperty.

### Expected Behavior
1) I expect that the get()/convention()/set() functions follow the same basic contract inherited from Property, just that this Property contains a Map or a Collection. E.g: I'd expect set(Map) to completely replace the Map configured via convention(Map).

2) I expect that the put()/putAll() and add()/addAll() act as append operators on the current Map or Collection returned by the get() function. E.g: I'd expect put("k1", "v1") to a) augment values configured via convention(Map) or if set(Map) had been called for it to b) augment the values configured via set(Map)

### Current Behavior
My expectations from 1) and 2b) are working as expected, but my expectation 2a) behaves as if I called: `convention(Map), empty(), put("k1", "v1")` and my convention values have been discarded.

This behavior is inconsistent.

### Context
I'm working on putting together some custom Gradle distributions targeting Gradle 7+ (like what's described here: https://docs.gradle.org/current/userguide/organizing_gradle_projects.html#sec:custom_gradle_distribution ).

For this custom Gradle distribution, I'm trying to augment the static default property values which I have configured via a gradle.properties file and included in the root of my distribution with some dynamically generated "default" property values that I configure via an `init.d/ci.init.gradle` script which I'm also including in the distribution.

This has not been an easy task. The init scripts are not doing basic things that I've grown to expect from working in the settings.gradle and build.gradle scripts. It feels like everywhere I turn I need to dig through the Gradle source files to figure out another "work-around" to get things going again.

I would like this experience to be better, so I'll be filing 5 tickets with hope that we can to improve this situation.

This is ticket 3 of 5

### Steps to Reproduce 
```
    def "adds single entries to convention using put"() {
        given:
        property.convention(['k1': 'v1'])
        property.put('k2', 'v2')
        property.put('k3', 'v3')

        expect:
        assertValueIs(['k1': 'v1', 'k2': 'v2', 'k3': 'v3'])
    }
```

### Your Environment
Gradle 7.2

**Comment by ljacomet** [comment]
This is really a documentation issue. A `convention` only is available if the user does not attempt to change anything on the provider.
So the `put/putAll` and `add/addAll` methods should clarify that they will indeed clear the convention.

As a build author, if you want to provide default values that remain present when the build user appends / replace some, do not use a convention but instead do a `set`.

**Comment by steffenyount** [comment]
As a Java developer:

1) I expect map-like-object's put/putAll methods to always be additive, like I have grown to expect from all Java java.lang.Map implementations ever:
https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#put-K-V-
https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#putAll-java.util.Map-

2) I also expect collection-like-object's add/addAll methods to always be additive, like I have grown to expect from all Java java.lang.Collection implementations ever:
https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#add-E-
https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#addAll-java.util.Collection-

3) It feels really wonky/inconsistent for the Gradle interface to be intentionally going against that pattern, and insisting that these methods should only be destructive in the case when MapProperty.get()/AbstractCollectionProperty.get() would return a map/collection based on the convention(Map/Collection) value, but then also insisting that these methods should be additive in the case when MapProperty.get()/AbstractCollectionProperty.get() would return a map based on the set(Map/Collection) value.

Sure the documentation could be improved to better explain that Gradle behaves in this wonky way, but it would be much better for Gradle to be boring (and not wonky) and for Gradle to not trip up it's users with a mess of little quirks/gotchas like this one that then require a roundtrip through the documentation to figure out.

Cons:
I expect that making this change will affect very few existing build scripts. Since the convention() values are currently being discarded, there's no reason for implementers of existing scripts to ever be calling the convention() method on these MapProperty and AbstractCollectionProperty objects before calling their respective put/putAll/add/addAll methods.

Pros:
I expect that making this change will be beneficial because it allows the convention() method to actually be used to supply default values from init scripts, plugins, etc which can then either be replaced wholesale via the set(Map/Collection) or updated/augmented individually via the put/putAll/add/addAll methods by the consumers of those init scripts, plugins, etc.

Note: Please see my PR ^^^ above for reference

**Comment by ljacomet** [comment]
The team agrees that this is an issue to discuss and for which we need a solution. However the currently offered PR is too big of a change to integrate at this time.

**Comment by jbartok** [comment]
Relates:
* #13364

**Comment by melix** [comment]
I share @chrisdennis point of view. My plugins often declare _convention_ values which are reasonable, but for lists and sets, it's extremely annoying the convention value is discarded as soon as `add` is called. It forces the plugin to expose methods used to compute the default values, so that the user can call them manually in order to avoid loosing them.

**Comment by lptr** [comment]
This is a blocker for the provider API migration right now. We can't convert list properties because of this problem.

**Comment by steffenyount** [comment]
> This is a blocker for the provider API migration right now. We can't convert list properties because of this problem.

@lptr I've updated my PR which remedies this issue. I fixed some merge conflicts that had come up. Please feel free to merge it.

Thanks.

**Comment by steffenyount** [comment]
Hi @abstratt,

Note:
1. I've rebased my PR branch.
2. If my original PR seems too invasive or complicated, please let me know. I have put together a less invasive patch that achieves the the same outcomes. You can find it [here](https://github.com/steffenyount/gradle/tree/append_operators_are_discarding_convention_values_V2).  I'm be happy to submit that alternate PR for for review. Please let me know if that is something you'd be interested in.

Best,
-Steffen

**Comment by abstratt** [comment]
@steffenyount We are considering exposing mutation API that operate on the convention/default value (so you can augment the existing convention), just like we have `add`/`addAll` or `put`/`putAll` to augment the explicit value.

Would that satisfy the use case you had, which I am reading as: "I need to augment (and not wholesale replace) existing *conventions/defaults*"?

*Note that I understand that is different from "I need to augment (and not wholesale replace) the actual value of a property, no matter whether the actual value is defined by the convention (by a plugin or shared build logic) or explicitly (by shared or local build logic)", which is more powerful, and which is what your proposed change promoted.*

**Comment by abstratt** [comment]
We are proposing an opt-in accessor (currently a method named `value()`) so people can update map/collection properties according to the semantics expected here. We do not think we have the freedom of changing the current behavior the existing incremental update methods in `MapProperty`/`HasMultipleValues`, which may eventually be deprecated.

The design spec is [here](https://docs.google.com/document/d/1KHCCzNeNAIkpedH9fUcNUe_vgZAWHxCODzxVW_7NzLw/edit).

You can see some snippets and test cases showing the new API action [here](https://github.com/gradle/gradle/pull/26564/files#diff-5591eb2c8933f9ed3fe2ab8d0f7f1d578aa1864597a0ae3aadc06e0e2a258a7b).

@steffenyount @melix @chrisdennis and others, your feedback is very appreciated.

**Comment by chrisdennis** [comment]
I'm generally much more comfortable reviewing things like this in their natural form (not through a spec document, at least not in isolation). Is the contents of #26564 the normative form? Am I okay to review that PR directly and only refer to the spec document for added context?

**Comment by steffenyount** [comment]
Hi @abstratt,

It looks like both the older and the newer proposals enable my use-case, but they are a bit cumbersome and don't address the big UX problem I've also tried to describe in this ticket: I want Gradle to be more standard, less weird, less tricky for the casual Gradle user to get right!

When a class/interface has Map in its name (like MapProperty) and includes Java's familiar Map functions: put(), putAll(), I don't want those functions to have Gradle's current nuanced and surprising behavior that's equivalent to:

when the MapProperty has only been specified via convention:
    then put() and putAll() are equivalent to calling "clear() + put()" or "clear() + putAll()" on a Java Map
    else put() and putAll() are equivalent to calling "put()" or "putAll()" on a Java Map

When a class/interface has List or Set  in its name (like ListProperty or SetProperty) and includes Java's familiar Collection functions: add(), addAll(), I don't want those functions to have Gradle's current nuanced and surprising behavior that's equivalent to:

when the ListProperty or SetProperty has only been specified via convention:
    then add() and addAll() are equivalent to calling "clear() + add()" or "clear() + addAll()" on a Java Collection
    else add() and addAll() are equivalent to calling "add()" or "addAll()" on a Java Collection

I think your newer proposal is better, in that the user of the new additive functions via the value() accessor isn't burdened with the extra cognitive load required to track the convention/value state of the MapProperty, ListProperty, or SetProperty.

That said the MapProperty, ListProperty, and SetProperty classes still include their put()/putAll()/add()/addAll() functions, where they still look like a Map/List/Set "duck", but then still don't really "quack" like a Map/List/Set "duck". If you also deprecate+remove these functions from the MapProperty, ListProperty, and SetProperty interfaces leaving the functionality accessible only through the new value() accessor then you also fix this UX problem since, it will be clearer to the user that the MapProperty, ListProperty, and SetProperty classes are not Map/List/Set "ducks".

But please, I'd like to suggest that you re-consider the assessment that "you don't have the freedom to change the current behavior" of the put()/putAll()/add()/addAll() functions:

As I tried to point out in the "Expected Behavior"/"Current Behavior" blocks of my initial ticket write up, and also in the later "cons:" section of my first response above, there is only a very small discrepancy between the existing behavior and the desired behavior for existing build scripts. The following 2 step migration path should be relatively straight forward to implement, and would allow the suggested behavior to be introduced on the existing functions at the next major release.

Step 1: In the next point release, deprecate or eliminate the very limited set of call sites where an implicit "empty()" call is being added before put()/putAll()/add()/addAll() calls and then require that the implicit "empty()" call is added there explicitly.

Basically, this would imply adding a new runtime warning or making it a runtime error to call the put()/putAll()/add()/addAll() functions on a MapProperty, ListProperty, or SetProperty when a non-empty convention has been specified and neither the empty() or the set() function has subsequently been called on that property.

Step 2: In the next major release (and possibly earlier under a feature flag), allow or reintroduce support for the previously (in step 1) banned call sites where an implicit "empty()" call would have been added, but then update the implementation so that it doesn't apply the implicit "empty()" call functionality anywhere.

**Comment by abstratt** [comment]
> Is the contents of #26564 the normative form? Am I okay to review that PR directly and only refer to the spec document for added context?

@chrisdennis  That PR is a spike/prototype/playing ground for trying out the ideas discussed in the spec. It is not a shape that is meant to be merged. Also, there will be times where the spec is updated and the PR lags behind, or the other way around, some experimentation is shared there before the spec is updated. If all you read is the PR, I'd stick to reading only the test case and API changes.

**Comment by abstratt** [comment]
@steffenyount We can and should deprecate and either remove or redefine the semantics of the existing incremental update methods to work the way we want.  No disagreement there. Considering the current proposal, `MapProperty.put*(...)` could eventually be reimplemented as being the same as invoking `MapProperty.value().put*(...)`.  Since the current proposal does not change the behavior of the existing API, we can allow ourselves some time to confirm the new API serves the use cases we have in a proper way, and then later decide how to migrate usages of the existing API to the new API. 

However, we would like to provide a more explicit way of configuring the convention, or the explicit value (instead of relying on whether an explicit value was set or not (or was cleared via `set(null)` or not). We intend to do that by exposing accessors for the explicit value and the convention value for those callers that care about the difference.

**Comment by abstratt** [comment]
Please track #24773 for documentation on the programming model proposed by this issue.

**Comment by abstratt** [comment]
Reopening since the (incubating) API we added in 8.7 has been removed in 8.8.

**Comment by lptr** [comment]
This is not a requirement for the provider API migration as migrated code should not yet have conventions.
